numpy/ma/core.py

```diff
@@ -4572,21 +4590,25 @@ def sum(self, axis=None, dtype=None, out=None):
         newmask = _check_mask_axis(_mask, axis)
         # No explicit output
         if out is None:
-            result = self.filled(0).sum(axis, dtype=dtype)
+            result = self.filled(0).sum(axis, dtype=dtype, keepdims=keepdims)
             rndim = getattr(result, 'ndim', 0)
             if rndim:
                 result = result.view(type(self))
-                result.__setmask__(newmask)
+                if (newmask is nomask):
+                    result.__setmask__(nomask)
+                else:
+                    result.__setmask__(newmask.reshape(result.shape))
             elif newmask:
                 result = masked
             return result
         # Explicit output
-        result = self.filled(0).sum(axis, dtype=dtype, out=out)
+        result = self.filled(0).sum(axis, dtype=dtype, out=out,
+                                    keepdims=keepdims)
         if isinstance(out, MaskedArray):
             outmask = getattr(out, '_mask', nomask)
             if (outmask is nomask):
                 outmask = out._mask = make_mask_none(out.shape)
-            outmask.flat = newmask
+            outmask[()] = newmask.reshape(result.shape)
```

Comment:
In the end I don't think I can replace `.flat =` with `[...] =`.

Using `.flat` is currently necessary for interopreability of ma and matrices. See https://github.com/numpy/numpy/pull/4585#issuecomment-40175006, https://github.com/numpy/numpy/pull/4615.

Eg, if you sum a 3x3 masked-matrix along the 2nd axis, the resulting data shape is `(3,1)`, but the mask shape becomes `(3,)` (since mask is an ndarray). `[...] = ` fails when trying to assign a `(3,)` shape to a `(3,1)` shape, but `.flat = ` doesn't. 

If we really want to get rid of `a.flat = b` here, something like this might be a replacement:

    a[...] = b.reshape(a.shape) if not np.isscalar(b) else b

but that seems a bit ugly.